Teach the importer to import any clang vector type as SIMDn<Scalar>.

stephentyrone opened the pull request:
Instead of having a known set of vector types, check to see if the element type conforms to SIMDScalar; if it does, see if we have a SIMDn defined with the right number of elements. If both are satisfied, import the vector type as that Swift type.

By making this change, we gain the ability to import vector types that aren't defined in terms of the Darwin simd module, which lets us use C API with vector types on other platforms. It also lets us
import *every* vector type that Swift can represent, rather than the small subset that are currently hardcoded.

Resovles <rdar://problem/34170174>

stephentyrone committed:
Teach the importer to import any vector type as SIMDN<Scalar>.

Instead of having a known set of vector types, check to see if the
element type conforms to SIMDScalar; if it does, see if we have a
SIMDN defined with the right number of elements. If both are satisfied,
import the vector type as that Swift type.

By making this change, we gain the ability to import vector types
that aren't defined in terms of the Darwin simd module, which lets
us use C API with vector types on other platforms. It also lets us
import *every* vector type that Swift can represent, rather than the
small subset that are currently hardcoded.

stephentyrone commented:
@rjmccall I'm probably doing a whole bunch of things wrong here, and I still need to implement the reverse mapping, but this is functional enough to get some feedback on while I work on the rest.

stephentyrone commented:
@swift-ci please smoke test

rjmccall commented:
Are these different types from what we use to import these as?

rjmccall commented on lib/ClangImporter/ImportType.cpp:
```diff
@@ -510,36 +510,32 @@ namespace {
+        name.append(std::to_string(count));
+        if (auto vector = Impl.getNamedSwiftType(Impl.getStdlibModule(), name)) {
+          if (auto unbound = vector->getAs<UnboundGenericType>()) {
+            ArrayRef<Type> arg(element);
```
I'm not sure this is okay memory-wise, but you should be able to just pass `{ arg }`.

rjmccall commented on lib/ClangImporter/ImportType.cpp:
```diff
@@ -510,36 +510,32 @@ namespace {
+                                              OptionalTypeKind::OTK_None);
+      unsigned count = type->getNumElements();
+      // Import vector-of-one as the element type.
+      if (count == 1) { return element; }
```
Hmm.  Does this always have the same layout as it would in C?

stephentyrone replied:
Up to manually-specified alignment, I think so. (How do we handle int32 typedefs that are over/under-aligned for that matter?)

rjmccall replied:
We probably ignore the alignment, which is probably a bug. :)

stephentyrone replied:
As long as it's consistently a bug, the same as other imported types =)

rjmccall commented on lib/ClangImporter/ImportType.cpp:
```diff
@@ -510,36 +510,32 @@ namespace {
+                                              ImportTypeKind::Abstract,
+                                              false,
+                                              Bridgeability::Full,
+                                              OptionalTypeKind::OTK_None);
```
I think you want `Bridgeability::None` here, although it probably doesn't actually matter.

Please add a `/* ... */` comment explaining the `false` argument.

stephentyrone commented:
@rjmccall we used to implement them as types that are typedefs of these types.

rjmccall commented:
Okay, that shouldn't cause source-compatibility problems, then.

stephentyrone committed:
Increased test coverage for increased SIMD types that we can import.

Includes some minor cleanup from review. Also eliminates the old
simd_sans_simd test, since we can now import all of these types even when the simd module isn't imported.

stephentyrone commented:
@swift-ci please test

stephentyrone commented:
@swift-ci please test source compatibility

stephentyrone commented:
@rjmccall Are you OK with me merging this assuming the compatibility test results are not unexpected?

rjmccall commented:
Yeah, go ahead and merge.
